### PR TITLE
Move the registry to in-memory cache

### LINKED ISSUE #6387: Move the plugin registry to stateless
# Description

As part of the issue that pretends to remove the stateful of the Wazuh dashboard plugins, we need to replace the way the data of the plugin's registry is stored so this doesn't use a file in the filesystem of the application.

Currently, this data is stored in the `wazuh-registry.json` file located at `<Wazuh Dashboard>/data/wazuh/config/wazuh-registry.json.

The data related to the API hosts will be stored in an in-memory cache managed instead. Any action about the API host configuration should update the registry data of the host.

# Tasks

- [x] Analyze the stored data in the registry file https://github.com/wazuh/wazuh-dashboard-plugins/issues/6387#issuecomment-1966213961
- [x] Propose the new store system https://github.com/wazuh/wazuh-dashboard-plugins/issues/6387#issuecomment-1980866235
- [x] Implement the new store system based on in-memory-cache
  - [x] Implement the cache on the `ManageHosts` service that will store information about the registry
  - [x] Remove the `UpdateRegistry` file
  - [x] Initializes the cache with the API hosts at start
  - [x] Move and remove the logic to update the registry from the frontend side to the backend side

### PR DESCRIPTION
**Pull request by Desvelao**
### Description
This pull request moves the registry to an in-memory cache and removes some logic done by the frontend side that should do the backend side.

#### Changes
- Move the registry management to an in-memory cache managed by ManageHosts
  service.
- Initialize the registry data on the start
  - Update the registry on adding, editing, or deleting an API host
- Move the logic to check if the run_as is enabled to the ManageHosts
  instead of cacheAPIUserAllowRunAs
- Update the backend types of the core plugin
- Remove the process to create/update the registry file when the plugin starts
- Remove the `Installation date` from the About view
- Remove the cacheAPIUserAllowRunAs of the ManageHosts service to use
  the registry data
- Remove the cacheAPIUserAllowRunAs property of Manage hosts
- Remove the UpdateRegistry service
- Remove API endpoint PUT /hosts/update-hostname/{id}
  - Remove API endpoint (route and controller)
  - Remove API request of frontend side
- Remove logic to create the registry file on start
- Remove API endpoint and usage of GET /hosts/remove-orphan-entries
  - Remove API endpoint (route and controller)
  - Remove API request of frontend side
- Remove unused API endpoint GET /api/timestamp
 
### Issues Resolved
#6387

### Evidence
image

### Test
# TEST

With an API connection using an with `run_as` capability enabled with `run_as` disabled, edit the API connection setting `run_as` as `true`, save. The `Run as` column should display the `run_as` is enabled with a `check`

Context: @yenienserrano 

# 6387-registry-in-memory-cache

Legend:
:black_circle:: none
:green_circle:: pass
:yellow_circle:: warning
:red_circle:: fail
:white_circle:: not applicable

## UI

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to About, it only should display the version and revision of the application. No reference to installation date. | :black_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:black_circle: Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to About, it only should display the version and revision of the application. No reference to installation date.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

# 6387-registry-in-memory-cache

Legend:
:black_circle:: none
:green_circle:: pass
:yellow_circle:: warning
:red_circle:: fail
:white_circle:: not applicable

## UI

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: | :black_circle: | :black_circle: |
| Go to About, it only should display the version and revision of the application. No reference to installation date. | :black_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:black_circle: Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:black_circle: Go to About, it only should display the version and revision of the application. No reference to installation date.</summary>

Chrome - :black_circle:

Firefox - :black_circle:

Safari - :black_circle:

</details>

### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by Desvelao** [comment]
# TEST

With an API connection using an with run_as` capability enabled with `run_as` disabled, edit the API connection setting `run_as` as `true`, save. The `Run as` column should display the `run_as` is enabled with a `check`

Context: @yenienserrano

File: plugins/main/server/controllers/wazuh-api.ts [hunk 1]
```diff
@@ -87,6 +84,7 @@ export class WazuhApiCtrl {
           idHost,
         );
       }
+      console.log('HELLO2');
```
Comment (yenienserrano) [inline plugins/main/server/controllers/wazuh-api.ts:87]: Perhaps this can be removed

**Comment by yenienserrano** [comment]
when I delete an API I get an error message but if I press the refresh button the API seems to have been deleted correctly because it does not appear in the table.

image

image

**Review by Tostti — CHANGES_REQUESTED** [review]
## UI

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :green_circle:  | :black_circle: |
| Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table. | :black_circle: | :green_circle:  | :black_circle: |
| Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: |:green_circle:  | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table. | :black_circle: |:green_circle:  | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table. | :black_circle: | :green_circle: | :black_circle: |
| Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table. | :black_circle: | :green_circle:  | :black_circle: |
| Go to About, it only should display the version and revision of the application. No reference to installation date. | :black_circle: | :green_circle: | :black_circle: |
| Changing run as status | :black_circle: | :red_circle: | :black_circle: |
| Removing an API | :black_circle: | :red_circle: | :black_circle: |

**Details**
<details>
<summary>:green_circle: Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image


Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image
image

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image

image


Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to About, it only should display the version and revision of the application. No reference to installation date.</summary>

Chrome - :black_circle:

Firefox - :green_circle:
image

Safari - :black_circle:

</details>


<details>
<summary>🔴  Changing run as status</summary>

Chrome - :black_circle:

Firefox - :red_circle:

Although the change is applied, it is not shown in the table
image
image
image


Safari - :black_circle:

</details>


<details>
<summary>🔴  Removing an API</summary>

Chrome - :black_circle:

Firefox - :red_circle:

If an API is removed (online or offline), this error is shown and the table have to be manually refreshed to show the change
image


Safari - :black_circle:

</details>

**Comment by yenienserrano** [comment]
when I edit an API I get an error message but if I press the update button the API seems to have been edited correctly. 

image
image

File: plugins/main/server/controllers/wazuh-api.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Desvelao) [reply plugins/main/server/controllers/wazuh-api.ts]: Removed here: cab4219a27a63e44fd0e77755f8009cd41b46179

Comment (Desvelao) [reply plugins/main/server/controllers/wazuh-api.ts]: Removed here: cab4219a27a63e44fd0e77755f8009cd41b46179

**Comment by Desvelao** [comment]
# Changes
- Fixed an error deleting the API host
- Fixed an error when editing an API connection changing the ID. Related to https://github.com/wazuh/wazuh-dashboard-plugins/pull/6481#issuecomment-2004503682
- Fixed an error when enabling the `Run as` and this doesn't display as enabled in the API connection table: Related to https://github.com/wazuh/wazuh-dashboard-plugins/pull/6481#pullrequestreview-1942801334

**Comment by asteriscos** [comment]
## UI

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table. | :green_circle: | :black_circle: | :black_circle: |
| Go to About, it only should display the version and revision of the application. No reference to installation date. | :green_circle: | :black_circle: | :black_circle: |

**Details**
<details>
<summary>:green_circle: Go to Server APIs, and add an online API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :green_circle:

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and edit an offline API host that can be connected. The data about the cluster name or manager name should appear in the table.</summary>

Chrome - :green_circle:

image

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and edit an online API host that can not be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :green_circle:

image

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and edit the API connection that can be connected. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :green_circle:

image

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, change the cluster name or manager name of an offline API connection, and click on the Refresh button. The new data about the cluster name or manager name should appear in the table.</summary>

Chrome - :green_circle:

image

image

Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to Server APIs, and add an offline API host that can be connected. The data about the cluster name or manager name should not appear in the table.</summary>

Chrome - :green_circle:

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

<details>
<summary>:green_circle: Go to About, it only should display the version and revision of the application. No reference to installation date.</summary>

Chrome - :green_circle:

image


Firefox - :black_circle:

Safari - :black_circle:

</details>

**Review by asteriscos — APPROVED** [review]

**Pull request merged**